### PR TITLE
feat: add vanities

### LINKED ISSUE #175: Implement Content Vanity API
# Content Vanity API

## v0.6.0 Proposal

```python
from posit import connect

client = connect.Client()
content = client.content.get(...)

vanity = content.vanity = ...
del content.vanity

vanity.destroy()
```

### PR DESCRIPTION
**Pull request by tdstein**
Adds vanities support.

A new mixin pattern is introduced, which adds the `vanity` attributes to the `ContentItem` class. This abstraction is still a bit leaky since the endpoint is hardcoded to 'v1/content' . But I find this design helpful since it collocates everything regarding vanities into a single file. In addition, the Content class has started to get a little unwieldy due to its size.

The use of TypedDict, Unpack, Required, and NotRequired from `typing_extensions` is also introduced. Thanks to https://blog.changs.co.uk/typeddicts-are-better-than-you-think.html, I discovered this pattern provides a much more succinct way to define pass-through arguments. It also greatly simplifies the verbosity of overload methods. That isn't shown here, but I have a test on another branch.

Resolves #175

### CODE REVIEW

**Comment by tdstein** [comment]
Sorry for the churn on this. While writing out the cookbook recipes, I discovered a few improvements. This should be good to go now.

**Comment by tdstein** [comment]
Ok, I'm pretty sure I'm actually done now :)

File: src/posit/connect/vanities.py [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+        after_destroy: Optional[AfterDestroyCallback] = None,
+        **kwargs: Unpack[VanityAttributes],
+    ):
+        """Initialize a Vanity.
+
+        Parameters
+        ----------
+        params : ResourceParameters
+        after_destroy : AfterDestroyCallback, optional
+            Called after the Vanity is successfully destroyed, by default None
```
Comment (toph-allen) [inline src/posit/connect/vanities.py:68]: This is interesting to me; what are its use cases?

File: src/posit/connect/vanities.py [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+
+class VanityMixin(Resource):
+    """Mixin class to add a vanity attribute to a resource."""
+
+    class HasGuid(TypedDict):
+        """Has a guid."""
+
+        guid: Required[str]
+
+    def __init__(self, /, params: ResourceParameters, **kwargs: Unpack[HasGuid]):
```
Comment (toph-allen) [inline src/posit/connect/vanities.py:128]: What does `/` mean?

File: src/posit/connect/vanities.py [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+        return [Vanity(self.params, **result) for result in results]
+
+
+class VanityMixin(Resource):
+    """Mixin class to add a vanity attribute to a resource."""
+
+    class HasGuid(TypedDict):
+        """Has a guid."""
+
+        guid: Required[str]
```
Comment (toph-allen) [inline src/posit/connect/vanities.py:126]: Could you explain how this works? Is the internal class definition here is mostly for type-related stuff? (I see it used in the `__init__` method below)

File: tests/posit/connect/test_vanities.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    @responses.activate
+    def test_destroy_calls_after_destroy_callback(self):
+        content_guid = "8ce6eaca-60af-4c2f-93a0-f5f3cddf5ee5"
+        base_url = "http://connect.example/__api__"
+        endpoint = f"{base_url}/v1/content/{content_guid}/vanity"
+        responses.delete(endpoint)
+
+        session = requests.Session()
+        url = Url(base_url)
+        after_destroy = Mock()
```
Comment (toph-allen) [inline tests/posit/connect/test_vanities.py:38]: Aha — is `after_destroy` a parameter so that it can be mocked?

**Review by toph-allen — APPROVED** [review]

File: src/posit/connect/vanities.py (hunk 1 above)
Comment (tdstein) [reply src/posit/connect/vanities.py:68]: It's currently used on L145 below (`self._vanity._after_destroy = self.reset_vanity`), which says, "after the vanity is destroyed, call reset_vanity". It's a way to control the state of VanityMixin without having to interact with it directly from Vanity.

File: src/posit/connect/vanities.py (hunk 3 above)
Comment (tdstein) [reply src/posit/connect/vanities.py:126]: Yep, it is strictly for the type checker. It makes this class easier to work with in unit tests.

File: src/posit/connect/vanities.py (hunk 2 above)
Comment (tdstein) [reply src/posit/connect/vanities.py:128]: https://docs.python.org/3/whatsnew/3.8.html#positional-only-parameters

Although, I no longer need this. It's cruft left over from implementation.

```suggestion
    def __init__(self, params: ResourceParameters, **kwargs: Unpack[HasGuid]):
```

**Pull request merged**